sql/parser: split into more separate packages

knz opened the pull request:
First commit from #19901 - will rebase when that merges.

tl;dr this patch separates items from the `sql/parser` package into
the following:

- new package `sem/builtins`: the built-in function *definitions*

- new package `sem/transforms`: `ContainsWindow`,
  `IsAggregateVisitor`, `ExprTransformContext` (not
  `NormalizeVisitor`, that's still entangled with `parser`)

- new package `sem/tests`: tests that need both the parser and the
  built-in function definitions, to break the dependency cycle.

Details follow.

The following `parser` definitions remain in the `parser` package for
now but are moved to their own files:

- `FunctionClass` and constants, `Builtin` and methods (`builtins.go` -> `builtin.go`)
- `AggregateFunc` (`aggregate_builtins.go` -> `aggregate_funs.go`)
- `ValueGenerator` (`generator_builtins.go` -> `generators.go`)
- `IndexedRow`, `WindowFrame`, `WindowFunc` (`window_builtins.go` -> `window_funs.go`)
- `intPow` (`builtins.go` -> `eval.go`)
- `pickFromTuple` (`builtins.go` -> `eval.go`)
- `mockNameTypes` (`type_check.go` -> `testutils.go`)
- `init`, (`builtins.go` -> `all_builtins.go`, then moved below)
- `collectBuiltins` (`window_builtins.go` -> `all_builtins.go`, then moved below)

- `TestCastToCollatedString` (`eval_test.go` -> `collatedstring_test.go`)
- `TestEvalComparisonExprCaching` (`eval_test.go` -> `compare_test.go`)
- `TestSimilarEscape`, `Benchmark*Like*` (`eval_test.go` -> `like_test.go`)
- `TestClusterTimestampConversion` (`eval_test.go` -> `timestamp_test.go`)

This patch then extracts code into the following separate modules:

- `canBeInArray` -> `IsValidArrayElementType` in `sql/sem/types`

- `ContainsWindowVisitor` -> `window.go` in `sql/sem/transform`
- `IsAggregateVisitor` -> `aggregates.go` in `sql/sem/transform`
- `expr_transform.go` -> `sql/sem/transform`

- `all_builtins.go`, `builtins.go` -> `sql/sem/builtins`
- `aggregate_builtins.go` -> `sql/sem/builtins`
- `aggregate_builtins_test.go` -> `sql/sem/builtins`
- `window_builtins.go` -> `sql/sem/builtins`
- `pg_builtins.go` -> `sql/sem/builtins`
- `EmptyDTable()` -> `sql/sem/builtins` (needs `arrayValueGenerator`, a built-in)

- `TestHelpFunctions` (`help_test.go` -> `sql/sem/builtins/help_test.go`)

The following tests are then in turn extracted into a new `sem/tests`
package, because they have a dependency on both `parser` and the new
`sem/builtins` package (they depend on the SQL built-in functions
being defined):

- `TestTypeCheck`, `TestTypeCheckError` -> (`type_check_test.go` -> `sql/sem/tests/type_check_test.go`)
- all tests from `format_test.go`
- all tests from `function_name_test.go`
- all tests from `normalize_test.go`
- all tests from `parse_test.go`
- all tests from `timeconv_test.go`
- remaining tests in `eval_test.go`

This in turn required the following API exports in `parser`:

- the fields of `Builtin`. This in turn removes the
  need for its public accessors.
- `AddWithOverflow`
- `AppendToMaybeNullArray`
- `ConcatArrays`
- `ErrDivByZero`
- `ErrZeroModulus`
- `EvalContext.StmtTimestamp`
- `EvalContext.TxnTimestamp`
- `FixedReturnType`, `UnknownReturnType`, `IdentityReturnType`
- `FunDefs`
- `IntPow`
- `MockNameTypes`
- `NewFunctionDefinition`
- `NormalizeVisitor`
- `PickFromTuple`
- `PrependToMaybeNullArray`
- `RegexpCacheKey` and its `Pattern` interface method.
- `ReturnTyper`
- `SecondsInDay`
- `TypeList`

The new following new items were added:

- `TimestampDifference`, to provide visibility into
  `parser.timestampMinusBinOp` from `sem/builtins` and clarify its
  intent.
- `FmtAlwaysGroupExprs` to provide access to the `alwaysParens` format
  flag from `sem/tests`.

The following was deleted, because unnecessary:

- `isNaN()` - inlined in caller, simplifies and reduces overhead

knz committed:
parser: move basic lexical functions to a new package 'lex'

tl;dr: this patch moves the token and keyword definitions, the lexical
predicates (isIdentifier etc.) and the string / identifier escaping
functions into a new package `sql/lex`.

This is achieved as follows:

- the main functions from `encode.go` are exported, then `encode.go`
  is moved to `sql/lex`.

- the lexical predicates are extracted from `scan.go`, moved to
  `predicates.go`, exported, then moved to `sql/lex`.

- the keyword definitions are moved to `sql/lex` because they are
  needed by the lexical predicates and the encode functions. This is
  achieved by modifying `all_keywords.awk`, and
  `reserved_keywords.awk` to refer to the new `lex` package, and
  modifying `Makefile` accordingly. The generated files `keywords.go`
  and `reserved_keywords` are also moved accordingly.

- Also, since `keywords.go` uses the token constant definitions (the
  mapping of token name to numeric value used in the parser), these
  are moved to `lex` as well, by adding a new extraction rule in
  `parser/Makefile`. The generated `sql.go` is edited automatically to
  alias these token definitions from `lex`, instead of defining them
  in-place.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19922)
<!-- Reviewable:end -->

knz commented:
@jordanlewis this is not quite ready to review yet. I'll add a 2nd commit that removes the alias stubs. However in the meantime can you confirm that the commit message is closer to what you'd like to see for #19901?

knz committed:
sql/parser: split into more separate modules

tl;dr this patch separates items from the `sql/parser` package into
the following:

- new package `sem/builtins`: the built-in function *definitions*
- new package `sem/transforms`: `ContainsWindow`,
  `IsAggregateVisitor`, `ExprTransformContext` (not
  `NormalizeVisitor`)
- new package `sem/tests`: tests that need both the parser and the
  built-in function definitions, to break the dependency cycle.

Details follow.

The following `parser` definitions remain in the `parser` package for
now but are moved to their own files:

- `FunctionClass` and constants, `Builtin` and methods (`builtins.go` -> `builtin.go`)
- `AggregateFunc` (`aggregate_builtins.go` -> `aggregate_funs.go`)
- `ValueGenerator` (`generator_builtins.go` -> `generators.go`)
- `IndexedRow`, `WindowFrame`, `WindowFunc` (`window_builtins.go` -> `window_funs.go`)
- `intPow` (`builtins.go` -> `eval.go`)
- `pickFromTuple` (`builtins.go` -> `eval.go`)
- `mockNameTypes` (`type_check.go` -> `testutils.go`)
- `init`, (`builtins.go` -> `all_builtins.go`, then moved below)
- `collectBuiltins` (`window_builtins.go` -> `all_builtins.go`, then moved below)
- `TestCastToCollatedString` (`eval_test.go` -> `collatedstring_test.go`)
- `TestEvalComparisonExprCaching` (`eval_test.go` -> `compare_test.go`)
- `TestSimilarEscape`, `Benchmark*Like*` (`eval_test.go` -> `like_test.go`)
- `TestClusterTimestampConversion` (`eval_test.go` -> `timestamp_test.go`)

This patch then extracts code into the following separate modules:

- `canBeInArray` -> `IsValidArrayElementType` in `sql/sem/types`
- `ContainsWindowVisitor` -> `window.go` in `sql/sem/transform`
- `IsAggregateVisitor` -> `aggregates.go` in `sql/sem/transform`
- `expr_transform.go` -> `sql/sem/transform`
- `all_builtins.go`, `builtins.go` -> `sql/sem/builtins`
- `aggregate_builtins.go` -> `sql/sem/builtins`
- `aggregate_builtins_test.go` -> `sql/sem/builtins`
- `window_builtins.go` -> `sql/sem/builtins`
- `pg_builtins.go` -> `sql/sem/builtins`
- `EmptyDTable()` -> `sql/sem/builtins` (needs `arrayValueGenerator`, a built-in)
- `TestHelpFunctions` (`help_test.go` -> `sql/sem/builtins/help_test.go`)

The following tests are then in turn extracted into a new `sem/tests`
package, because they have a dependency on both `parser` and the new
`sem/builtins` package (they depend on the SQL built-in functions
being defined):

- `TestTypeCheck`, `TestTypeCheckError` -> (`type_check_test.go` ->
  `sql/sem/tests/type_check_test.go`)2
- all tests from `format_test.go`
- all tests from `function_name_test.go`
- all tests from `normalize_test.go`
- all tests from `parse_test.go`
- all tests from `timeconv_test.go`
- remaining tests in `eval_test.go`

This in turn required the following API exports in `parser`:

- the fields of `Builtin`. This in turn removes the
  need for its public accessors.
- `AddWithOverflow`
- `AppendToMaybeNullArray`
- `ConcatArrays`
- `ErrDivByZero`
- `ErrZeroModulus`
- `EvalContext.StmtTimestamp`
- `EvalContext.TxnTimestamp`
- `FixedReturnType`, `UnknownReturnType`, `IdentityReturnType`
- `FunDefs`
- `IntPow`
- `MockNameTypes`
- `NewFunctionDefinition`
- `NormalizeVisitor`
- `PickFromTuple`
- `PrependToMaybeNullArray`
- `RegexpCacheKey` and its `Pattern` interface method.
- `ReturnTyper`
- `SecondsInDay`
- `TypeList`

The new following new items were added:

- `TimestampDifference`, to provide visibility into
  `parser.timestampMinusBinOp` from `sem/builtins` and clarify its
  intent.
- `FmtAlwaysGroupExprs` to provide access to the `alwaysParens` format
  flag from `sem/tests`.

The following was deleted, because unnecessary:

- `isNaN()` - inlined in caller, simplifies and reduces overhead

knz committed:
sql/lex: simplify encodeEscapedSQLIdent.

Moving this function to its own `lex` package reveals
it is not used elsewhere.

Its only two callers always call it with the bareIdentifiers flag set,
so there is really no need for the parameter.

Simplify the function accordingly.

knz commented:
hold on to reviewing this, I'm splitting it up in smaller patches to make it more easy.

knz commented:
Superseded by #19941.
